semaphore/semaphore.go

```diff
@@ -0,0 +1,71 @@
+// Package sempaphore provides a simple implementation of a semaphore (hence the name, duh).
+// Semaphores may be usefull to enforce a upper bound on how much
+// concurrent units executes some algorithm that is very heavy on memory and/or CPU.
+//
+// One example is when you have an http server that can accept hundreds
+// of concurrent requests but you have some critical section of your code that
+// cant handle that much concurrency at once (but you don't want to serialize either).
+//
+// There are other ways to handle this too using channels, which approach to use will
+// be your judgement call. If you choose the semaphore here you got one.
+package semaphore
+
+import (
+	"fmt"
+	"context"
+)
+
+// S is a semaphore instance. Always use the New function
+// to create semaphores. Using a unproperly initialized
+// semaphore may cause mayhem on your code.
+type S chan struct{}
+
+// Release is used to release a previous call to S.Acquire
+type Release func()
+
+// New creates a new semaphore with the given size.
+// Passing 0 as size is a moronic programming mistake and will
+// result in a panic due to its moronicness.
+func New(size uint) S {
+	if size == 0 {
+		panic("semaphore.New:cant create a semaphore with size 0")
+	}
+	return S(make(chan struct{}, size))
+}
+
+// Acquire will acquire the semaphore, if the semaphore is
+// already full Acquire will wait until some other goroutine
+// releases the semaphore or until the given context timeouts.
+//
+// If the context timeouts or is cancelled a non nil error is returned.
+// On success a release function is returned, this function can be used
+// only to release the correspondent call to Acquire, calling the
+// returned Release function twice will lock your goroutine forever since
+// this is a very stupid thing to do (better than releasing the semaphore twice
+// I think, at least the bugged code is the one who will get screwed).
+//
+// Never calling release is also a terrible idea since this may cause
+// starvation to resources if the semaphore is used to provide controlled
+// access to some resource (usually an expensive one).
+func (s S) Acquire(ctx context.Context) (Release, error) {
+	select {
+		case s <- struct{}{}:
+			{
+				released := false
```

Comment:
theoretically, user still can call this function twice, concurrently, because there's a race in the closure `released` variable. 

What do you think of `release` doing nothing in the subsequent calls? The first call releases the resources, but others are no-op. This way we avoid the panic and we can also make it concurrent safe.

```go
select {
    case s <- struct{}{}:
        var rmu = &sync.Mutex{}
        var released bool
        return func() {
            rmu.Lock()
            defer rmu.Unlock()
            if released {
                return
            }
            released = true
            <-s
         }, nil
}
```
There's no cost in the mutex also, as release will only be called once in normal circumstances.

Reply:
I know you were aware, sorry because it was bad written. What I mean is, there's a race and maybe we can make something better because the current mechanism could not suffice. But now I take back regarding making it a no-op.. silently failing is very bad in this case.
Now I think you need to fix the race, but only to safely panic. It's just few lines of code and then you'll deliver a _correct_ mechanism to avoid releasing the resource twice *and* will notify the programmer of the mistake.

_I understand that the race exist only if the programmer makes a mistake, but as we return a release function and Go code is highly concurrent and adding that Go has defer that commonly could lead to double-call.. I dont see much impossible to occur_.